### DISCUSSION TITLE
How can I load Homebrew's Zsh Completions?

### DISCUSSION THREAD

**Discussion by metebyte** [z-shell/zi#158] [Q&A] [answered]
Hello,

As I mentioned on the title, Is there a way to load Homebrew's Zsh Completions located at `/usr/local/share/zsh/site-functions`? And Is it okay to load those completions along with ZI?

**Comment by ss-o** [comment]


**Reply by ss-o** [reply]
Hey @metebyte :wave:

I always prefer best practices and according to the official docs suggest:

```zsh
if type brew &>/dev/null
then
  FPATH="$(brew --prefix)/share/zsh/site-functions:${FPATH}"

  autoload -Uz compinit
  compinit
fi
```

If it's correct I can convert it to ZI syntax :octocat:

**Reply by ss-o** [reply]
🧙‍‍♀️ Here is my suggestion 🧙‍‍♀️

> It will also be put under full ZI control to avoid conflicts with other apps and prevents leaks and other pollution.

```zsh
zi has'brew' id-as'brew-completions' wait as'completion' lucid \
   atclone='print Installing Brew completions...; mkdir -p $ZPFX/brew_stuff; \
   command cp -f $(brew --prefix)/share/zsh/site-functions/^_* $ZPFX/brew_stuff; \
zi creinstall -q $(brew --prefix)/share/zsh/site-functions' \
   atload='fpath=( ${(u)fpath[@]:#$(brew --prefix)/share/zsh/*} ); fpath+=( $ZPFX/brew_stuff )' \
   atpull="%atclone" run-atpull for \
     z-shell/null
```

After this, you need to call `zicompinit`, and it has to be called at the right time, see:
[calling-compinit-with-turbo-mode](https://z.digitalclouds.dev/docs/guides/commands/#calling-compinit-with-turbo-mode)

Finally, I recommend setting `ZI[ZCOMPDUMP_PATH]` if completions are managed by ZI. More:
[customization#-customizing-paths](https://z.digitalclouds.dev/docs/guides/customization#-customizing-paths)

We can discuss the lines if needed 😵‍💫 It is important to understand how ZI works - and it's pretty easy 🤓 The **SECRET KEY 🗝️** is just to get the point 🙇‍‍♂️

**Reply by ss-o** [reply]
By the way, forgot to mention that these complex functions can be written as a [package](https://z.digitalclouds.dev/ecosystem/packages).

If there would be a group of complex commands then all can be written as a [meta-plugin](https://z.digitalclouds.dev/ecosystem/annexes/meta-plugins/).

The complex command above would look like this:

```shell
zi pack for brew-stuff
```

Additionally, packages can be written with different configurations and chosen when required:

```shell
zi pack'mac-fpath' for brew-stuff
zi pack'zi-zpfx' for brew-stuff
```

**Reply by metebyte** [reply]
> 🧙‍‍♀️ Here is my suggestion 🧙‍‍♀️
> 
> > It will also be put under full ZI control to avoid conflicts with other apps and prevents leaks and other pollution.
> 
> ```shell
> zi has'brew' id-as'brew-completions' wait as'completion' lucid \
>    atclone='print Installing Brew completions...; mkdir -p $ZPFX/brew_stuff; \
>    command cp -f $(brew --prefix)/share/zsh/site-functions/^_* $ZPFX/brew_stuff; \
> zi creinstall -q $(brew --prefix)/share/zsh/site-functions' \
>    atload='fpath=( ${(u)fpath[@]:#$(brew --prefix)/share/zsh*} ); fpath+=( $ZPFX/brew_stuff )' \
>    atpull="%atclone" run-atpull for \
>      z-shell/null
> ```
> 
> After this, you need to call `zicompinit`, and it has to be called at the right time, see: [calling-compinit-with-turbo-mode](https://z.digitalclouds.dev/docs/guides/commands/#calling-compinit-with-turbo-mode)
> 
> Finally, I recommend setting `ZI[ZCOMPDUMP_PATH]` if completions are managed by ZI. More: [customization#-customizing-paths](https://z.digitalclouds.dev/docs/guides/customization#-customizing-paths)
> 
> We can discuss the lines if needed 😵‍💫 It is important to understand how ZI works - and it's pretty easy 🤓 The **SECRET KEY 🗝️** is just to get the point 🙇‍‍♂️

This is exactly what I was looking for. Thanks.

**Reply by metebyte** [reply]
> By the way, forgot to mention that these complex functions can be written as a [package](https://z.digitalclouds.dev/ecosystem/packages).
> 
> If there would be a group of complex commands then all can be written as a [meta-plugin](https://z.digitalclouds.dev/ecosystem/annexes/meta-plugins/).
> 
> The complex command above would look like this:
> 
> ```shell
> zi pack for brew-stuff
> ```
> 
> Additionally, packages can be written with different configurations and chosen when required:
> 
> ```shell
> zi pack'mac-fpath' for brew-stuff
> zi pack'zi-zpfx' for brew-stuff
> ```

Is there a difference between this way and packaging way? I mean about performance or avoiding misconfiguration with other ZI stuff?

**Reply by ss-o** [reply]
A package is a set of instructions as it would be in `.zshrc` or written by you - no difference.
So you can have long and complex commands or clean and good-looking `.zshrc`. 😎

**Comment by metebyte** [comment, marked as answer]
So here's my first experience 😄 

After loading completions with the method you mentioned above, I got this error with git completions.

```
$ git com[TAB]
```

turned into this

```
$ git comgit com
_git:.:48: no such file or directory:
```

**Reply by metebyte** [reply]
I think It's not about the method or ZI. It's about git. Because I rolled back to my old config and uninstalled whole completions and rebuild it with zicompinit. Still getting the same error with `zsh-users/zsh-completions` during the git command completion.

So I wanted to give a try to ZI way.

```
zi ice lucid wait as'completion' blockf mv'git-completion.zsh -> _git'
zi snippet https://github.com/git/git/blob/master/contrib/completion/git-completion.zsh
```

Tried this way mentioned on the Wiki and still getting the same error.

**Reply by metebyte** [reply]
After uninstalling `fzf-tab` error disappeared. By the way I couldn't mark your reply as an answer because none of your replies have that button 😞

**Reply by ss-o** [reply]
You get errors only about `_git`, while there are multiple files. 🙇‍‍♂️ 💡
You do not trust the command? 👀 (Joke)

The problem is [git-completion](https://github.com/felipec/git-completion/), which had a bug when the plugin directory is a symlink.

Also, don't forget that completions are compiled and contained in memory and need to be re-compiled.

---

Have the below in `.zshrc`:

```shell
zi has'brew' wait lucid id-as'brew-comps' as'completion' atclone'mkdir -p $ZPFX/brew_stuff; \
  command cp -f $(brew --prefix)/share/zsh/site-functions/^_* $ZPFX/brew_stuff; \
    zi creinstall -q $(brew --prefix)/share/zsh/site-functions' atpull'%atclone' \
      atload'fpath=( ${(u)fpath[@]:#$ZPFX/share/zsh/*} ); fpath+=( $ZPFX/brew_stuff )' for \
        z-shell/null
```

Try to install the latest `git-completion`. Make sure that this is executed below the above one.

Try first:

```shell
zi wait lucid as'completion' blockf mv'*.zsh -> _git' for \
  felipec/git-completion
```

then: `zi report felipec/git-completion`

If you don't see anything you like and it still doesn't work then:

Do some cleaning:

```shell
zi cclear
zi cuninstall git-completion
zi delete felipec/git-completion
```
And replace with this:

```shell
zi wait lucid as'completion' blockf mv'*.zsh -> $ZPFX/brew_stuff/_git' \
  atload'zstyle ':completion:*:*:git:*' script $ZPFX/brew_stuff/_git' for \
    felipec/git-completion
```

> - Ice: [blockf](https://z.digitalclouds.dev/docs/getting_started/overview/#the-completion-management)

--- 

If all done and working then you can control it with ZI. Even commands are long (can be packed as mentioned), from know ZI will take care of your completion system.
Also it's it will be updated same way when run `zi update`.

**Reply by metebyte** [reply]
Isn't this output weird?

```
❯ zi report felipec/git-completion
Report for felipec/git-completion plugin
----------------------------------------

Completions:
_git [enabled]
_git [enabled]
```

**Reply by metebyte** [reply]
Tried some completions for git and it didn't break. So It's working atm. Thanks for the info!

**Reply by ss-o** [reply]
![z_git](https://user-images.githubusercontent.com/59910950/166614998-2c4e7381-0073-462f-8f84-23eb60aaaa9c.png)

More than welcome, try to practice from wiki, as it will be always improved according to ZI. :1st_place_medal: 

Place this in your `.zshrc`:

```zsh
rain-zi() { echo -e "\e[1;40m" ; clear ; while :; do echo $LINES $COLUMNS $(( $RANDOM % $COLUMNS)) $(( $RANDOM % 72 )) ;sleep 0.05; done|awk '{ letters="abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789@#$%^&*()"; c=$4;        letter=substr(letters,c,1);a[$3]=0;for (x in a) {o=a[x];a[x]=a[x]+1; printf "\033[%s;%sH\033[2;32m%s",o,x,letter; printf "\033[%s;%sH\033[1;37m%s\033[0;0H",a[x],x,letter;if (a[x] >= $1) { a[x]=0; } }}' }
```

Reload shell and run: `rain-zi` :tada: 

Have a great day! :octocat:

**Reply by ss-o** [reply]
@metebyte I also have some power on GitHub, know you can mark answer :octocat:

**Reply by metebyte** [reply]
Since the first and root reply of yours is deleted there isn't an option to mark 🤔

**Reply by ss-o** [reply]
This discussions got recently updated UI, need some work to be done. 
All organization profiles and GitHub getting updated heavily. :octocat:

**Reply by metebyte** [reply]
I see, anyway I'm so grateful for your help and to you for being that patient and helpful 🫡

**Reply by ss-o** [reply]
@metebyte,

Try `zi pack for brew-completions` :octocat:

**Reply by metebyte** [reply]
I will 👌 Btw Is GitHub Discussions only option for communicating?

**Reply by ss-o** [reply]
https://join.slack.com/t/z-shell/shared_invite/zt-16twpopd2-p08ROUeT2aGZ5njJwysawA

**Reply by ss-o** [reply]
> I will ok_hand Btw Is GitHub Discussions only option for communicating?

Which you prefer?

**Reply by metebyte** [reply]
> https://join.slack.com/t/z-shell/shared_invite/zt-16twpopd2-p08ROUeT2aGZ5njJwysawA

That is what I was looking for 🙃

**Comment by j9ac9k** [comment]
Hello z-shell devs, I'm having the same issue that the author is having, the fixes suggested here haven't worked for me, with what I am sure that's because of something else that is conflicting in my `.zshrc`

Here are the contents of my `.zshrc`

```zsh
# Fig pre block. Keep at the top of this file.
. "$HOME/.fig/shell/zshrc.pre.zsh"
export CLICOLOR=1

if [[ ! -f $HOME/.zi/bin/zi.zsh ]]; then
  print -P "%F{33}▓▒░ %F{160}Installing (%F{33}z-shell/zi%F{160})…%f"
  command mkdir -p "$HOME/.zi" && command chmod g-rwX "$HOME/.zi"
  command git clone -q --depth=1 --branch "main" https://github.com/z-shell/zi "$HOME/.zi/bin" && \
    print -P "%F{33}▓▒░ %F{34}Installation successful.%f%b" || \
    print -P "%F{160}▓▒░ The clone has failed.%f%b"
fi
source "$HOME/.zi/bin/zi.zsh"
autoload -Uz _zi
(( ${+_comps} )) && _comps[zi]=_zi
# examples here -> https://z.digitalclouds.dev/ecosystem/annexes
# zicompinit # <- https://z.digitalclouds.dev/docs/guides/commands

# zi has'brew' wait lucid id-as'brew-comps' as'completion' atclone'mkdir -p $ZPFX/brew_stuff; \
#   command cp -f $(brew --prefix)/share/zsh/site-functions/^_* $ZPFX/brew_stuff; \
#     zi creinstall -q $(brew --prefix)/share/zsh/site-functions' atpull'%atclone' \
#       atload'fpath=( ${(u)fpath[@]:#$ZPFX/share/zsh/*} ); fpath+=( $ZPFX/brew_stuff )' for \
#         z-shell/null

# theme
zi ice pick"async.zsh" src"pure.zsh"
zi light sindresorhus/pure

# fzf
zi ice lucid wait has'fzf'
zi light Aloxaf/fzf-tab

# Completions
zi pack for brew-completions

zi wait lucid for \
 atinit"ZI[COMPINIT_OPTS]=-C; zicompinit; zicdreplay" \
    z-shell/F-Sy-H \
 blockf \
    zsh-users/zsh-completions \
 atload"!_zsh_autosuggest_start" \
    zsh-users/zsh-autosuggestions

zi ice lucid wait as'completion' blockf has'pandoc'
zi light srijanshetty/zsh-pandoc-completion

zi ice lucid wait as'completion' blockf has'mpv'
zi snippet https://github.com/mpv-player/mpv/blob/master/etc/_mpv.zsh

# zi ice lucid wait as'completion' blockf mv'git-completion.zsh -> _git'
# zi snippet https://github.com/git/git/blob/master/contrib/completion/git-completion.zsh

zi wait lucid as'completion' blockf mv'*.zsh -> _git' for \
  felipec/git-completion

# zi wait lucid as'completion' blockf mv'*.zsh -> $ZPFX/brew_stuff/_git' \
#   atload'zstyle ':completion:*:*:git:*' script $ZPFX/brew_stuff/_git' for \
#     felipec/git-completion

zi ice lucid wait as'completion' blockf has'ffmpeg'
zi snippet https://github.com/zsh-users/zsh/blob/master/Completion/Unix/Command/_ffmpeg

zi ice lucid wait as'completion' blockf has'pip'
zi snippet "$SHELL_COMMON/zsh/completions/_pip"

zi light z-shell/z-a-bin-gem-node
zi pack for ls_colors

zi pack"bgn" git for pyenv
# pyenv virtualenv
eval "$(pyenv init --path)"
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"
# Created by `pipx` on 2022-05-07 05:17:06
export PATH="$PATH:/Users/ogi/.local/bin"
export EDITOR="subl"

# Fig post block. Keep at the bottom of this file.
. "$HOME/.fig/shell/zshrc.post.zsh"
```


The issue is when I go to a git repo and try and use any kind of tab completion I get this kind output:

```zsh
~/some/git/repo
❯ git rm doc<TAB>
_git:.:48: no such file or directory:
__git_zsh_bash_func:8: command not found: __git_aliased_command
~/some/git/repo
> 
```

Things do work if I remove this bit of my `.zshrc`

```zsh
zi wait lucid for \
 atinit"ZI[COMPINIT_OPTS]=-C; zicompinit; zicdreplay" \
    z-shell/F-Sy-H \
 blockf \
    zsh-users/zsh-completions \
 atload"!_zsh_autosuggest_start" \
    zsh-users/zsh-autosuggestions
```

but then I lose my pretty coloring 😄 

In case it helps, I'm on macOS 12.3.1, I'm using [fig](https://github.com/withfig/fig) with iTerm2.

Thanks for making an awesome product!

EDIT: similar issue here: https://github.com/zdharma-continuum/zinit/issues/124

**Reply by pskarma** [reply]
@j9ac9k,

Thanks for the detailed report. I am already splitting your `.zshrc` :construction_worker_man:

> @ss-o I'll do the magic search.

**Reply by ss-o** [reply]
This issue has been solved in a Slack Workspace:

---

[ss-o](https://github.com/ss-o) [19:22](https://z-shell.slack.com/archives/D03FG8LFE1Y/p1652638957278019)
Hey :wave:
you removed `felipec/git-completion` or `git/git` and left `zsh/users` completions?


[j9ac9k](https://github.com/j9ac9k) [19:26](https://z-shell.slack.com/archives/D03FG8LFE1Y/p1652639165324669)
i removed both :grimacing:

---
